Framework/Source/Graphics/Program/ProgramReflection.cpp

```diff
@@ -0,0 +1,1190 @@
+            uint32_t descOffset = getDescOffsetFromPath(pPath, max(1u, pType->getTotalArraySize()), category);
+            pVar = ReflectionVar::create(name, pType, index, descOffset, space);
+        }
+        else
+        {
+            pVar = ReflectionVar::create(name, pType, pSlangLayout->getOffset());
+        }
+        return pVar;
+    }
+
+    ReflectionVar::SharedPtr reflectTopLevelVariable(VariableLayoutReflection* pSlangLayout)
+    {
+        ReflectionPath path;
+        path.pVar = pSlangLayout;
+        return reflectVariable(pSlangLayout, &path);
+    }
+
+    static void storeShaderVariable(const ReflectionPath& path, SlangParameterCategory category, const std::string& name, ProgramReflection::VariableMap& varMap, ProgramReflection::VariableMap* pVarMapBySemantic, uint32_t count, uint32_t stride)
+    {
+        ProgramReflection::ShaderVariable var;
+        const auto& pTypeLayout = path.pVar->getTypeLayout();
+        var.type = getVariableType(pTypeLayout->getScalarType(), pTypeLayout->getRowCount(), pTypeLayout->getColumnCount());
+
+        uint32_t baseIndex = (uint32_t)getRegisterIndexFromPath(&path, category);
+        for(uint32_t i = 0 ; i < max(count, 1u) ; i++)
+        {
+            var.semanticIndex = baseIndex + (i*stride);
+            var.semanticName = path.pVar->getSemanticName();
+            if (count)
+            {
+                var.semanticName += '[' + std::to_string(i) + ']';
+            }
+            varMap[name] = var;
+            if (pVarMapBySemantic)
+            {
+                (*pVarMapBySemantic)[var.semanticName] = var;
+            }
+        }
+    }
+
+    static void reflectVaryingParameter(const ReflectionPath& path, const std::string& name, SlangParameterCategory category, ProgramReflection::VariableMap& varMap, ProgramReflection::VariableMap* pVarMapBySemantic = nullptr)
+    {
+        TypeLayoutReflection* pTypeLayout = path.pVar->getTypeLayout();
+        // Skip parameters that don't consume space in the given category
+        if (pTypeLayout->getSize(category) == 0) return;
+
+        TypeReflection::Kind kind = pTypeLayout->getKind();
+        // If this is a leaf node, store it
+        if ((kind == TypeReflection::Kind::Matrix) || (kind == TypeReflection::Kind::Vector) || (kind == TypeReflection::Kind::Scalar))
+        {
+            storeShaderVariable(path, category, name, varMap, pVarMapBySemantic, 0, 0);
+        }
+        else if (kind == TypeReflection::Kind::Array)
+        {
+            auto arrayKind = pTypeLayout->getElementTypeLayout()->getKind();
+            assert((arrayKind == TypeReflection::Kind::Matrix) || (arrayKind == TypeReflection::Kind::Vector) || (arrayKind == TypeReflection::Kind::Scalar));
+            uint32_t arraySize = (uint32_t)pTypeLayout->getTotalArrayElementCount();
+            uint32_t arrayStride = (uint32_t)pTypeLayout->getElementStride(SLANG_PARAMETER_CATEGORY_UNIFORM);
```

Comment:
This won't work with `SLANG_PARAMETER_CATEGORY_UNIFORM`, since a varying parameter will use zero bytes of uniform data (and thus will have a zero-byte stride). You need to be using `category` here, I think.